File: ruby/lib/oppex_sdk/internal/interrupt.rb
# frozen_string_literal: true

module Oppex
  module Internal
    # A one-way "stop waiting" signal.
    #
    # A sleeping Ruby thread cannot be interrupted safely the way a Java thread
    # can, so a backoff delay waits on this condition variable instead of calling
    # +sleep+. Closing the client signals it once, which wakes every waiting
    # retry immediately rather than after up to eight more seconds.
    #
    # Not public API.
    class Interrupt
      def initialize
        @mutex = Mutex.new
        @changed = ConditionVariable.new
        @signalled = false
      end

      # Signals every current and future waiter. Idempotent.
      def signal
        @mutex.synchronize do
          @signalled = true
          @changed.broadcast
        end
      end

      def signalled?
        @mutex.synchronize { @signalled }
      end

      # Waits up to +seconds+. Returns true when the wait was cut short by a
      # signal, false when the full duration elapsed.
      def wait(seconds)
        deadline = monotonic_now + seconds
        @mutex.synchronize do
          until @signalled
            remaining = deadline - monotonic_now
            return false if remaining <= 0

            @changed.wait(@mutex, remaining)
          end
          true
        end
      end

      private

      # A monotonic clock, so a system clock adjustment mid-backoff cannot turn a
      # half-second wait into a very long one.
      def monotonic_now
        Process.clock_gettime(Process::CLOCK_MONOTONIC)
      end
    end
  end
end

File: rust/src/internal/interrupt.rs
use std::sync::{Condvar, Mutex};
use std::time::Duration;

/// A one-way "stop waiting" signal.
///
/// A sleeping thread cannot be interrupted in Rust the way a Java thread can, so
/// a backoff delay waits on this condition variable instead of calling
/// [`std::thread::sleep`]. Closing the client signals it once, which wakes every
/// waiting retry immediately rather than after up to eight more seconds.
#[derive(Debug)]
pub(crate) struct Interrupt {
    signalled: Mutex<bool>,
    changed: Condvar,
}

impl Interrupt {
    pub(crate) fn new() -> Self {
        Self {
            signalled: Mutex::new(false),
            changed: Condvar::new(),
        }
    }

    /// Signals every current and future waiter. Idempotent.
    pub(crate) fn signal(&self) {
        // A poisoned lock means another thread panicked while holding it. The
        // flag itself is still consistent, so recover rather than propagate a
        // panic out of a shutdown path.
        let mut signalled = self
            .signalled
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        *signalled = true;
        self.changed.notify_all();
    }

    /// Waits up to `duration`. Returns `true` when the wait was cut short by a
    /// signal, `false` when the full duration elapsed.
    pub(crate) fn wait(&self, duration: Duration) -> bool {
        let signalled = self
            .signalled
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        let (signalled, _) = self
            .changed
            .wait_timeout_while(signalled, duration, |signalled| !*signalled)
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        *signalled
    }
}
